src/service/xhr-impl.js

```diff
@@ -112,8 +113,16 @@ export class Xhr {
    * @private
    */
   fetchAmpCors_(input, opt_init) {
+    const init = opt_init || {};
     input = this.getCorsUrl(this.win, input);
-    return this.fetch_(input, opt_init).then(response => {
+    // For some same origin requests, add AMP-Same-Origin: true header to allow
+    // publishers to validate that this request came from their own origin
+    // when Origin header is not going to be set by the browser.
+    if (!this.willSetOriginHeader_(input, init)) {
```

Comment:
Could we simplify this (origin + method + browser checking) to "if same origin then set AMP-Same-Origin header"? It's simpler, and it can also be surprising that not all AMP same origin XHRs contain the AMP-Same-Origin: true header (there are no security consequences, just weird).